plugins/Core-Extensions/Recommended/edu.toronto.cs.se.modelepedia.operator/src/edu/toronto/cs/se/modelepedia/operator/slice/Slice.java

```diff
@@ -74,20 +76,181 @@ public Input(Map<String, Model> inputsByName) {
+			equalFlag = false;
+		}
+		
+		return equalFlag;
+		
+				
+	}
+
+	// Returns the complete list of model elements that may be impacted
+	// by the model elements included in the original slicing criterion.
+	// By default, all model elements are assumed to be impacted by the slice.
+	public List<EObject> getImpactedElements(
+			List<EObject> unprocessed, List<EObject> unchecked) {
+		
+		List<EObject> changed = new ArrayList<EObject>();
+		Iterator<EObject> unprocessedIter = unprocessed.iterator();
+		
+		while (unprocessedIter.hasNext()) {
+			System.out.println("Elements to be processed: " + 
+					+ unprocessed.size());
+			EObject obj = unprocessedIter.next();
+
+			if (changed.contains(obj)){
+				// This element has already been checked for dependencies.
+				unprocessed.remove(obj);
+
+			} else {
+				unprocessed.remove(obj);
+				changed.add(obj);
+				System.out.println("Current element: " + obj);
+				
+				// Get all model elements affected by the impacted model element.
+				EObject uncheckedObj;
+				Iterator<EObject> uncheckedIter = unchecked.iterator();
+				while (uncheckedIter.hasNext()) {
+					uncheckedObj = uncheckedIter.next();
+					if (isModelElemImpactedBy(uncheckedObj, obj)) {
+						unprocessed.add(uncheckedObj);
+						uncheckedIter.remove();
+					}
+				}
+			}
+			
+			unprocessedIter = unprocessed.iterator();
+		}
+		
+		return changed;
+	}
+	
+	// Checks whether the first input model element is potentially
+	// impacted by the second second input model element.
+	public boolean isModelElemImpactedBy(EObject obj1, EObject obj2) {
+		return true;
+	}
+
+	protected ModelRel slice(ModelRel critRel, Model model, MID outputMID) throws MMINTException {
 		ModelRel sliceRel = critRel.getMetatype().createInstanceAndEndpoints(null, OUT_MODELREL, ECollections.newBasicEList(model), outputMID);
-		//TODO MMINT[MODELS17] The following is just an example of how to create a model element reference
-		EObject modelRootObj = model.getEMFInstanceRoot();
-		sliceRel.getModelEndpointRefs().get(0).createModelElementInstanceAndReference(modelRootObj, null);
 
+		List<EObject> criterionList = new ArrayList<>();
+		List<EObject> changed = new ArrayList<EObject>();
+		List<EObject> unprocessed = new ArrayList<EObject>();
+		List<EObject> unchecked = new ArrayList<EObject>();
+
+		// Get the input class diagram.
+		EObject root = model.getEMFInstanceRoot();
+		
+		// Add all elements in the class diagram to the unchecked list.
+		unchecked = getAllModelElements(model, root);
+
+		criterionList = getCriterionElements(critRel);
+
+		// Iterate through the unchecked list of class diagram elements to
+		// identify all those which are included in the input criterion.
+		unprocessed = matchCriterionToModelElements(unchecked, criterionList);
+		
+		// Iterate through the unprocessed list of impacted elements
+		// to identify all dependent elements that are also impacted.
+		changed = getImpactedElements(unprocessed, unchecked);
+
```

Comment:
The general idea is correct, but it's executed inefficiently, i.e. this will not scale to big or maybe average sized models (which is one thing we'll try to do in the paper) because of the number of nested loops and because the model files are loaded in memory multiple times.

getAllModelElements + getCriterionElements + matchCriterionToModelElements will not gain you any new information that you don't already have. That is, unprocessed is equals to running getEMFInstanceObject() on each ModelElementReference in critRel, and unchecked is everything else (which you probably don't really need to keep track).

getImpactedElements is where the interesting things happen.